{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Feature_Engineering.ipynb",
      "private_outputs": true,
      "provenance": [],
      "collapsed_sections": [],
      "machine_shape": "hm",
      "authorship_tag": "ABX9TyMDe4nfkV54Gp9dfTu6z63Y",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/GeorgeQiao/prostate_cancer/blob/master/Feature_Engineering.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "F2y5MLvnsLFS"
      },
      "source": [
        "'''\n",
        "Predicting Features from the Pathology Report for Prostate Cancer.\n",
        "\n",
        "'''\n",
        "\n",
        "# import statements\n",
        "import os\n",
        "import warnings\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import sklearn as sk\n",
        "from sklearn.model_selection import KFold, cross_val_score\n",
        "from sklearn.feature_selection import VarianceThreshold, SelectKBest, chi2, SelectFromModel\n",
        "from sklearn.preprocessing import StandardScaler\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "from sklearn import decomposition\n",
        "from imblearn.combine import SMOTEENN\n",
        "from imblearn.over_sampling import SMOTE\n",
        "from collections import Counter\n",
        "from sklearn.decomposition import PCA, KernelPCA\n",
        "from sklearn.discriminant_analysis import LinearDiscriminantAnalysis\n",
        "from sklearn.svm import SVC\n",
        "from sklearn.ensemble import RandomForestClassifier\n",
        "from sklearn.neighbors import KNeighborsClassifier\n",
        "from sklearn.naive_bayes import GaussianNB\n",
        "import matplotlib.pyplot as plt\n",
        "from mpl_toolkits import mplot3d\n",
        "from sklearn.ensemble import IsolationForest\n",
        "#from feature_processing import do_t_stage, do_gleason, do_t_recur\n",
        "\n",
        "\n",
        "# set module variables \n",
        "pd.set_option(\"expand_frame_repr\", False)\n",
        "pd.set_option(\"max_columns\", 9)\n",
        "np.random.seed(0)\n",
        "warnings.filterwarnings(\"ignore\")\n",
        "\n",
        "\n",
        "\n",
        "modes = ['show', 'partial', 'hide']\n",
        "mode = modes[1]\n",
        "\n",
        "# filepaths to data downloaded from cBioPortal\n",
        "filenames=[\"TCGA-FPKM-DEG.xlsx\"]\n",
        "\n",
        "# import the gene data (contains gene data to use as x-features while predicting)\n",
        "data = []\n",
        "for i, filename in enumerate(filenames):\n",
        "    TCGA_PRAD = pd.read_excel(\"TCGA-FPKM-DEG.xlsx\", header =None)\n",
        "    TCGA_PRAD = TCGA_PRAD.T\n",
        "    a = TCGA_PRAD\n",
        "    a = a.T\n",
        "    header_1 = a.iloc[0]\n",
        "    a = a[1:]\n",
        "    a.columns = header_1\n",
        "    a = a.reset_index()\n",
        "    TCGA_processed = a\n",
        "\n",
        "    TCGA_processed = TCGA_processed.rename(columns={np.nan: 'samples'})\n",
        "        # data is tab separated\n",
        "    data.append(TCGA_processed)\n",
        "\n",
        "    if mode == 'show':\n",
        "        print('FILENAME: {0} -\\n'.format(filenames[i]), data[i].head())\n",
        "\n",
        "# Transpose the gene data and add the target features to it\n",
        "t_data = []\n",
        "for i, d in enumerate(data):\n",
        "    t_data.append(d.fillna(d.mean()))\n",
        "    #missing = clinical_data.index.difference(d.transpose().index).tolist()\n",
        "\n",
        "print(t_data)\n",
        "# change TStage and Recurrence to integer values\n",
        "for i, d in enumerate(t_data):\n",
        "    t_data[i] = t_data[i][pd.notnull(t_data[i]['group'])]\n",
        "    enc = LabelEncoder()\n",
        "    # fit encoder to TStage \n",
        "    enc.fit(t_data[i]['group'].tolist())\n",
        "    t_data[i]['group'] = enc.transform(t_data[i]['group'].tolist())\n",
        "    tstage_mapping = dict(zip(enc.transform(enc.classes_), enc.classes_))\n",
        "    # fit encoder to Recurrence\n",
        "\n",
        "print(\"\\t\\tRisk-Level\")\n",
        "\n",
        "kbest = SelectKBest(score_func=chi2, k=350)\n",
        "scaler = StandardScaler()\n",
        "fs_data = []\n",
        "for i, d in enumerate(t_data):\n",
        "    print(\"\\nFILENAME: {}\".format(filenames[i]))\n",
        "    t_rows = list(d.index)\n",
        "    t_columns = d.columns[3:]\n",
        "    print(t_columns)\n",
        "    # K-best\n",
        "    print(d)\n",
        "\n",
        "    selector = kbest.fit(d.iloc[:, 3:], d.iloc[:, 2])\n",
        "    t_columns = t_columns[selector.get_support()]\n",
        "    fs_data.append(pd.DataFrame(selector.transform(t_data[i].iloc[:, 3:]), columns = t_columns, index=t_rows))\n",
        "    if mode == 'show':\n",
        "        print(\"Selecting k best features -\\n\", fs_data[i].head())\n",
        "    # Scale \n",
        "    t_columns = fs_data[i].columns\n",
        "    fs_data[i] = pd.DataFrame(scaler.fit_transform(fs_data[i]), columns=t_columns, index=t_rows)\n",
        "    if mode == 'show':\n",
        "        print(\"Scaling data -\\n\", fs_data[i].head())\n",
        "    # Select from RF\n",
        "    classifier = RandomForestClassifier(n_estimators=1)\n",
        "    classifier = classifier.fit(fs_data[i], d['group'])\n",
        "    selector = SelectFromModel(classifier, prefit=True)\n",
        "    t_columns = t_columns[selector.get_support()]\n",
        "    fs_data[i] = pd.DataFrame(selector.transform(fs_data[i]), columns=t_columns, index=t_rows)\n",
        "    fs_data[i]['group'] = d['group']\n",
        "    if mode == 'show':\n",
        "        print(\"Selecting data from RF model -\\n\", fs_data[i].head())\n",
        "    print(\"Shape after feature selection: {}\".format(fs_data[i].shape), end=\"\\n\\n\")\n",
        "# RESAMPLING the data - SMOTEENN\n",
        "balanced_data = [[] for _ in range(2)]\n",
        "for i, d in enumerate(fs_data):\n",
        "    sme = SMOTEENN(random_state=42, smote=SMOTE(random_state=42, k_neighbors=3))\n",
        "    x, y = sme.fit_resample(fs_data[i], t_data[i]['group'])\n",
        "    # x are the features and y are the targets\n",
        "    balanced_data[i].append(x)\n",
        "    balanced_data[i].append(y)\n",
        "    if mode == 'show':\n",
        "        print(\"FILENAME: {}\".format(filenames[i]), Counter(balanced_data[i][1]))\n",
        "# DIMENSIONALITY REDUCTION\n",
        "# Kernel PCA and LDA (can be toggled on or off)\n",
        "pca = True\n",
        "pca_dim = 31\n",
        "lda = True\n",
        "lda_dim = 5\n",
        "if pca or lda:\n",
        "    dr_data = []\n",
        "    for i in range(len(filenames)):\n",
        "        print(\"\\nFILENAME: {}\".format(filenames[i]))\n",
        "        if pca:\n",
        "            decomposer = KernelPCA(n_components=pca_dim, kernel='rbf', gamma=0.05, degree=7)\n",
        "            dr_data.append(decomposer.fit_transform(balanced_data[i][0]))\n",
        "            print(\"Shape and type after PCA: \", dr_data[i].shape, type(dr_data[i]))\n",
        "        if lda:\n",
        "            decomposer = LinearDiscriminantAnalysis(n_components=lda_dim, solver='eigen')\n",
        "            dr_data[i] = decomposer.fit_transform(dr_data[i], balanced_data[i][1])\n",
        "            print(\"Shape and type after LDA: \", dr_data[i].shape, type(dr_data[i]))\n",
        "else:\n",
        "    dr_data.append(balanced_data[0][0])\n",
        "    dr_data.append(balanced_data[1][0])\n",
        "# CLASSIFICATION\n",
        "splits = 7\n",
        "seed = 7\n",
        "kfold = KFold(n_splits=splits, random_state=seed, shuffle=True)\n",
        "results = {'SVM': [],\n",
        "            'RF': [],\n",
        "            'KNN': [],\n",
        "            'NB': []\n",
        "            }\n",
        "for i, d in enumerate(dr_data):\n",
        "    # SVM\n",
        "    res = []\n",
        "    classifier = SVC(gamma='auto')\n",
        "    results['SVM'].append(cross_val_score(classifier, pd.DataFrame(dr_data[i]), balanced_data[i][1], cv=kfold))\n",
        "    results['SVM'][i] = results['SVM'][i].mean()\n",
        "    # RF\n",
        "    classifier = RandomForestClassifier(n_estimators=100, max_depth=2, random_state=7, criterion='gini')\n",
        "    results['RF'].append(cross_val_score(classifier, pd.DataFrame(dr_data[i]), balanced_data[i][1], cv=kfold))\n",
        "    results['RF'][i] = results['RF'][i].mean()\n",
        "    # KNN\n",
        "    knn = KNeighborsClassifier(n_neighbors=5)\n",
        "    results['KNN'].append(cross_val_score(knn, pd.DataFrame(dr_data[i]), balanced_data[i][1], cv=kfold))\n",
        "    results['KNN'][i] = results['KNN'][i].mean()\n",
        "    # NB\n",
        "    nb = GaussianNB()\n",
        "    results['NB'].append(cross_val_score(nb, pd.DataFrame(dr_data[i]), balanced_data[i][1], cv=kfold))\n",
        "    results['NB'][i] = results['NB'][i].mean()\n",
        "print(\"\\nFinal Results for datasets: {0} -\".format(filenames[0]))\n",
        "print(results)\n",
        "# PLOTTING\n",
        "# PCA\n",
        "pca = PCA(n_components = 3)\n",
        "x_pca = pca.fit_transform(balanced_data[0][0])\n",
        "fig = plt.figure(figsize=(10, 6))\n",
        "plt.suptitle(\"3-D plot for resampled data using dimesnionality reduction (group)\\n\\n\")\n",
        "ax = fig.add_subplot(121, projection='3d')\n",
        "ax.set_title(\"PCA\\n\\n\")\n",
        "ax.view_init(elev=15,azim=66)\n",
        "for i in range(len(balanced_data[0][1])):\n",
        "    if balanced_data[0][1][i] == 0:\n",
        "        t2a = ax.scatter(x_pca[i][0], x_pca[i][1], x_pca[i][2], c='y', label=balanced_data[0][1][i])\n",
        "    elif balanced_data[0][1][i] == 1:\n",
        "        t2b = ax.scatter(x_pca[i][0], x_pca[i][1], x_pca[i][2], c='g', label=balanced_data[0][1][i])\n",
        "plt.legend((t2a, t2b),\n",
        "    ('0', '1'),\n",
        "    scatterpoints=1,\n",
        "    loc='upper right',\n",
        "    ncol=1,\n",
        "    fontsize=10)\n",
        "# PCA + LDA\n",
        "pca = PCA(n_components = 10)\n",
        "x_pca = pca.fit_transform(balanced_data[0][0])\n",
        "lda = LinearDiscriminantAnalysis(n_components = 3)\n",
        "x_lda = lda.fit_transform(x_pca, balanced_data[0][1])\n",
        "#print(x_lda)\n",
        "ax = fig.add_subplot(122, projection='3d')\n",
        "plt.title(\"PCA & LDA\\n\\n\")\n",
        "ax.view_init(elev=-79,azim=-7)\n",
        "# for i in range(len(balanced_data[0][1])):\n",
        "#     if balanced_data[0][1][i] == 0:\n",
        "#         t2a = ax.scatter(x_lda[i][0], x_lda[i][1], x_lda[i][2], c='y', label=balanced_data[0][1][i])\n",
        "#     elif balanced_data[0][1][i] == 1:\n",
        "#         t2b = ax.scatter(x_lda[i][0], x_lda[i][1], x_lda[i][2], c='g', label=balanced_data[0][1][i])\n",
        "# plt.legend((t2a, t2b),\n",
        "#             ('0', '1'),\n",
        "#             scatterpoints=1,\n",
        "#             loc='upper right',\n",
        "#             ncol=1,\n",
        "#             fontsize=10)\n",
        "\n",
        "# Plot the final results\n",
        "# DNA is the methylation hm450 file\n",
        "dna_res = {'SVM': [],\n",
        "        'RF': [],\n",
        "        'NB': [],\n",
        "        'KNN': []\n",
        "    }\n",
        "# RNA is the rna sequencing file\n",
        "rna_res = {'SVM': [],\n",
        "        'RF': [],\n",
        "        'NB': [],\n",
        "        'KNN': []\n",
        "    }\n",
        "r_results=r_risk_results=results\n",
        "for i in dna_res.keys():\n",
        "    dna_res[i].append(r_results[i][0])\n",
        "\n",
        "fig = plt.figure(figsize=(13, 7))\n",
        "fig.suptitle(\"Final Results\")\n",
        "# DNA FILE\n",
        "ax = fig.add_subplot(121)\n",
        "N = len(dna_res['SVM'])\n",
        "ind = np.arange(N)\n",
        "w = 0.11\n",
        "svm = ax.bar(ind, dna_res['SVM'], w, bottom=0, alpha=0.8, color='g')\n",
        "rf = ax.bar(ind+w, dna_res['RF'], w, bottom=0, alpha=0.8, color='r')\n",
        "nb = ax.bar(ind+2*w, dna_res['NB'], w, bottom=0, alpha=0.8, color='b')\n",
        "knn = ax.bar(ind+3*w, dna_res['KNN'], w, bottom=0, alpha=0.8, color='y')\n",
        "ax.set_title(\"Results using the classifiers for different target features on DNA File\")\n",
        "ax.set_xticks(ind+3*w/2)\n",
        "ax.set_xticklabels(('Risk'))\n",
        "ax.legend((svm[0], rf[0], nb[0], knn[0]), ('SVM', 'RF', 'NB', 'KNN'), loc='lower right')\n",
        "ax.yaxis.set_units('Percentages (%)')\n",
        "# RNA FILE\n",
        "\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}